enhancements/update/upgrades-blocking-on-ack.md

```diff
@@ -0,0 +1,138 @@
+---
+title: upgrade-acknowledgment-gate
+authors:
+  - "@jottofar"
+reviewers:
+  - "@LalatenduMohanty"
+  - "@bparees"
+  - "@sdodson"
+  - "@csrwng"
+  - "@relyt0925"
+  - "@jeremyeder"
+approvers:
+  - TBD 
+creation-date: 2021-07-08
+last-updated: 2021-07-08
+status: implementable
+---
+
+# Upgrade Acknowledgment Gate
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+CVO currently blocks minor level upgrades when overrides are set, a resource deletion is pending, or an operator has set its upgradeable condition to false.
+This enhancement adds a generic upgrade blocking mechanism for blocking minor level upgrades.
+This generic upgrade blocker can consist of one or more items each of which will block a minor level upgrade.
+Each generic upgrade blocker item will be associated with some well-defined task to be completed by an administrator.
+Once the task has been completed the administrator will acknowledge completion and its associated item will no longer block minor level upgrades.
+
+## Motivation
+
+OCP 4.9 uses kubernetes v1.22 which is removing a large number of [APIs][k8s-1.22-removed-apis].
+We suspect that many customer clusters are still using these APIs through helm charts, operators, generic workloads, administrative tools, etc.
+We want to prevent customers from upgrading from 4.8.z to 4.9.z and encountering missing API related breakage.
+
+### Goals
+
+* Acknowledgment requires the explicit update of a cluster resource by an administrator.
+* For the initial release in 4.8.z, administrator acknowledgment of API removal will be required before a minor level upgrade to 4.9 is allowed.
+* Upgrade blocking mechanism must be generic such that it can be repurposed in the future.
+* It must be clear what an administrator is acknowledging.
+* Administrators can bypass the upgrade blocking mechanism by using the existing CVO "force" implementation.
+* The UI will clearly indicate when this upgrade blocking mechanism is blocking upgrades, why, and what needs to be done to clear the block.
+* This upgrade blocking mechanism must be compatible with SD, ACM, IKS/ROKS, Hive, and Hypershift.
+
+### Non-Goals
+
+## Proposal
+
+CVO's existing framework for blocking minor level upgrades will be leveraged.
+The existing framework periodically checks upgradeable conditions.
+A new [upgradeable check method][upgradeableCheck-interface] will be added to the existing framework to implement this upgrade blocking mechanism.
+Use of this existing framework also allows this new upgrade blocking mechanism to be bypassed by using the `force` upgrade option.
+
+The new method will check a newly defined configmap for one or more predefined keys.
```

Comment:
Should these be predefined in code or could this be payload driven, ie: we add configmaps like this and each one is an independent ack.
```
apiVersion: v1
data:
  acknowledged: "false"
  message: |
    Kubernetes 1.22 and thus OpenShift 4.9 remove several APIs previously deprecated, please review the referenced documentation
  referenceURI: https://docs.openshift.com/yatta-yatta
kind: ConfigMap
metadata:
  annotations:
    clusterversion.openshift.io/admin-ack-request: "true"
    release.openshift.io/create-only: "true"
  name: admin-ack-request-kube-122-api-removals
  namespace: openshift-cluster-version
```

Reply:
This may be getting too close to needing a CRD / API.

Reply:
I like defining the key and message as you have above rather than hard-coding in CVO. I did interpret `we define a particular key name that the admin must create in the configmap with a value of “true” ` literally and write up as such thinking perhaps we didn't want to make it too easy for the admin to do the ack.  But we can still have a configmap to define the keys - its message and even it's valid version info - and  the separate configmap I defined that would be used by the admin to accomplish the ack'ing and checked by CVO. Plus we may want to keep the admin out of the "key config" configmap and limit to the less busy configmap used only for acking.

/cc @bparees

Reply:
> we didn't want to make it too easy for the admin to do the ack. But we can still have a configmap to define the keys - its message and even it's valid version info - and the separate configmap I defined that would be used by the admin to accomplish the ack'ing and checked by CVO. 

this direction matches a conversation scott and i were having on slack.  I definitely don't want it to be "too easy" but having the gates be defined by a separate config resource seems reasonable to me.

Reply:
Trying to nail down configmap(s), take a look...

I'm thinking there is one configmap per "gate" that defines the gate - name, message, applicable versions. CVO reads these configmaps to know what gates it needs to check, applicable version(s), message to generate if blocking. Then there's another configmap that the admin will be directed to to add the appropriate annotation, i.e. to do the ACK, for all defined gates. CVO is checking this configmap to ensure all defined gates have been entered or CVO will block minor upgrade.

For example below are two gates each defined by a configmap. Based on this config CVO will be checking `upgrade-ack` configmap (defined at bottom) for annotations `ack-4.9-api-removal-upgrade-approved: "true"` and `ack-another-ack: "true"` 
otherwise minor upgrades will be blocked.

```
apiVersion: v1
data:
  ackAnnoName: "ack-4.9-api-removal-upgrade-approved"
  message: |
    Kubernetes 1.22 and thus OpenShift 4.9 remove several APIs previously deprecated, please review the referenced documentation
  referenceURI: https://docs.openshift.com/yatta-yatta
  versions: "4.9"
kind: ConfigMap
metadata:
  name: admin-ack-request-kube-122-api-removals
  namespace: openshift-cluster-version
---
apiVersion: v1
data:
  ackAnnoName: "ack-another-ack"
  message: |
    Here's another ack, please review the referenced documentation...
  versions: "4.9,4.10"
kind: ConfigMap
metadata:
  name: admin-ack-another
  namespace: openshift-cluster-version
```
I'm thinking the admin does not have to create this configmap but merely update it with the appropriate anno which will be explained in detail in the doc's the admin is directed to:
```
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
  name: upgrade-ack
  namespace: openshift-cluster-version
```

Reply:
the version numbers should probably indicate "you need this ack if you are on this version" rather than "you need this ack before you can go to this version".  in case we ever allow skipping versions somehow.

Reply:
Actually, wouldn't my one-per-gate example configmaps conflict because they both contain the same keys, e.g. ackAnnoName, etc?

For that reason, and because I think it just makes more sense, we should have one configuration config map that defines everything about each "gate" as in:
```
apiVersion: v1
metadata:
  name: admin-ack-config
  namespace: openshift-cluster-version
data:
   upgrade.ack.configuration: |
      upgrade.ack.gates {
         gate: [
            {
               name: "ack-4.9-api-removal-upgrade-approved"
               valid.in.versions: "4.9"
               message: |
                  Kubernetes 1.22 and thus OpenShift 4.9 remove several APIs previously deprecated, please review the referenced documentation
               referenceURI: https://docs.openshift.com/yatta-yatta
            }
            {
               name: "ack-foo"
               valid.in.versions: "4.9,4.10"
               message: |
                  Foo bar, please review the referenced documentation
               referenceURI: https://docs.openshift.com/foo-bar
            }
         ]
      }

CVO reads 0 to n gates and for each read, if it applies to the current OCP version, uses the `name` field to look it up in the `upgrade-ack` configmap and if not found to have been "ACK'ed" blocks minor upgrades.
If we decided a gate no longer applied, for example `ack-foo` becomes OBE in 4.10.9, we could just remove it from the above configmap.

Reply:
After discussing with @sdodson and @bparees we decided against additional configmap(s) due to added complexity of maintaining and consuming and instead go with simplicity of hard-coded key names.